File: frontend/src/utils/password.ts
/**
 * Shared password-strength policy for the frontend.
 *
 * Mirrors the backend rules in `backend/app/core/password.py`. Keep the two
 * in sync: any change here must be reflected there (and vice versa).
 */

export const PASSWORD_MIN_LENGTH = 12

export interface PasswordRuleStatus {
  label: string
  met: boolean
}

const RULES: { label: string; test: (password: string) => boolean }[] = [
  { label: `At least ${PASSWORD_MIN_LENGTH} characters`, test: (pw) => pw.length >= PASSWORD_MIN_LENGTH },
  { label: 'One lowercase letter', test: (pw) => /[a-z]/.test(pw) },
  { label: 'One uppercase letter', test: (pw) => /[A-Z]/.test(pw) },
  { label: 'One number', test: (pw) => /[0-9]/.test(pw) },
  { label: 'One special character', test: (pw) => /[^A-Za-z0-9]/.test(pw) },
]

/** Per-rule pass/fail status, for rendering a live checklist. */
export function evaluatePassword(password: string): PasswordRuleStatus[] {
  return RULES.map((rule) => ({ label: rule.label, met: rule.test(password) }))
}

/** True when the password satisfies every rule in the policy. */
export function isPasswordValid(password: string): boolean {
  return RULES.every((rule) => rule.test(password))
}

File: backend/app/core/password.py
"""
Arbiter shared password-strength policy.

Single source of truth for the password rules. Every flow that sets a
password (register, change-password, reset-password, accept-invite) runs
through here so the requirements stay consistent across the app. The
frontend mirrors these rules in ``components/PasswordRequirements.tsx``.
Keep the two in sync.
"""

from __future__ import annotations

PASSWORD_MIN_LENGTH = 12


def password_problems(password: str) -> list[str]:
    """Return human-readable descriptions of every unmet rule.

    An empty list means the password satisfies the policy.
    """
    problems: list[str] = []
    if len(password) < PASSWORD_MIN_LENGTH:
        problems.append(f"at least {PASSWORD_MIN_LENGTH} characters")
    if not any(c.islower() for c in password):
        problems.append("a lowercase letter")
    if not any(c.isupper() for c in password):
        problems.append("an uppercase letter")
    if not any(c.isdigit() for c in password):
        problems.append("a number")
    if not any(not c.isalnum() for c in password):
        problems.append("a special character")
    return problems


def validate_password_strength(password: str) -> str:
    """Pydantic/endpoint validator: raise ``ValueError`` listing unmet rules.

    Returns the password unchanged when it satisfies the policy so this can
    be used directly as a Pydantic ``field_validator`` body.
    """
    problems = password_problems(password)
    if problems:
        raise ValueError("Password must contain " + ", ".join(problems))
    return password
